latch/registry/table.py

```diff
@@ -0,0 +1,275 @@
+        ...
+
+
+# todo(ayush): "DeleteRecordUpdate", "UpsertColumnUpdate", "DeleteColumnUpdate"
+@dataclass
+class UpsertRecordUpdate(TableUpdate):
+    name: str
+    data: Dict[str, Any]
+
+    async def execute(self, session: gql.Client):
+        record_id = await session.execute(
+            gql.gql(
+                """
+                mutation CatalogMultiUpsertSamples($tableId: BigInt!, $names: [String!]!) {
+                    catalogMultiUpsertSamples(
+                        input: { argExperimentId: $tableId, argNames: $names }
+                    ) {
+                        bigInts
+                        clientMutationId
+                    }
+                }
+                """
+            ),
+            {
+                "tableId": self.table.id,
+                "names": [self.name],
+            },
+        )["catalogMultiUpsertSamples"]["bigInts"][0]
+
+        for column in self.table.get_columns():
+            literal = None
+            if column["key"] not in self.data:
+                if not column["type"]["allowEmpty"]:
+                    raise ValueError(
+                        f"unable to upsert record {self.name} as column"
+                        f" {column['key']} is missing a value"
+                    )
+            else:
+                literal = to_registry_literal(
+                    self.data[column["key"]],
+                    column["type"]["type"],
+                )
+
+            await session.execute(
+                """
+                mutation CatalogMultiUpsertColumnDatum(
+                    $data: [JSON]!
+                    $key: String!
+                    $sampleIds: [BigInt!]!
+                ) {
+                    catalogMultiUpsertColumnSampleData(
+                        input: { argSampleIds: $sampleIds, argKey: $key, argData: $data }
+                    ) {
+                        clientMutationId
+                    }
+                }
+                """,
+                {
+                    "data": [literal],
+                    "key": column["key"],
+                    "sampleIds": [record_id],
+                },
+            )
+
+
+class TableUpdater:
+    def __init__(self, table: Table):
+        self.table = table
+        self.updates: List[TableUpdate] = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, tb):
+        if type is not None or value is not None or tb is not None:
+            return False
+        self.commit()
+
+    def upsert_record(self, record_name: str, column_data: Dict[str, Any]):
+        self.updates.append(
+            UpsertRecordUpdate(
+                self.table,
+                record_name,
+                column_data,
+            )
+        )
+
+    def commit(self):
+        async def helper():
+            async with gql.Client(transport=get_transport()) as session:
+                for update in self.updates:
```

Comment:
these need to be merged into a single query, otherwise there is no reason to have the updater thing

at the very least leave a todo